html/captive-portal/lib/captiveportal/PacketFence/Controller/TLSProfile.pm

```diff
@@ -0,0 +1,180 @@
+        service             => $request->param_encoded("service"),
+        profile_list        => $Config{'pki'}{'profile'},
+        template            => 'pki.html',
+        provisioner         => $provisioner,
+        username            => $username,
+        );
+}
+sub cert_p12 : Path('/eap-profile.html') : Args(0) {
+    my ( $self, $c ) = @_;
+    my $cert_data = $c->response->body($c->stash->{'cert_content'});
+    #my $headers = $c->response->headers;
+    #$headers->content_type('application/x-pkcs12');
+    my $certname = $c->stash->{'certificate_cn'} . "p12";
+    #$headers->header( 'Content-Disposition', "attachment; filename=\"$certname\"" );
+    open FH, "> $cert_dir/$certname" or die $!;
+    print FH "$cert_data\n";
+}
+
+
+sub get_cert : Private {
+    use bytes;
+    my ($self,$c,) = @_;
+    my $stash = $c->stash;
+    my $uri = $Config{'pki'}{'uri'};
+    my $username = $Config{'pki'}{'username'};
+    my $password = $Config{'pki'}{'password'};
+    my $email = $stash->{'certificate_email'};
+    my $dot1x_username = $stash->{'certificate_cn'};
+    my $organisation = $Config{'pki'}{'organisation'};
+    my $state = $Config{'pki'}{'state'};
+    my $profile = $Config{'pki'}{'profile'};
+    my $country = $Config{'pki'}{'country'};
+    my $certpwd = $stash->{'certificate_pwd'};
+    my $response = '';
+    my $curl = WWW::Curl::Easy->new; #$self->curl($function);
+    my $request = "username=$username&password=$password&cn=$dot1x_username&mail=$email&organisation=$organisation&st=$state&country=$country&profile=$profile&pwd=$certpwd";
+    my $response_body = '';
+    $curl->setopt(CURLOPT_POSTFIELDSIZE,length($request));
+    $curl->setopt(CURLOPT_POSTFIELDS, $request);
+    $curl->setopt(CURLOPT_WRITEDATA, \$response_body);
+    $curl->setopt(CURLOPT_HEADER, 0);
+    $curl->setopt(CURLOPT_DNS_USE_GLOBAL_CACHE, 0);
+    $curl->setopt(CURLOPT_NOSIGNAL, 1);
+    $curl->setopt(CURLOPT_URL, $uri);
+  
+    # Starts the actual request
+    my $curl_return_code = $curl->perform;
+
+    $response = $response_body;
+    $c->stash(
+        cert_content    => $response_body,
+    );
+}
+ 
+sub cert : Local {
+    my ($self,$c,) = @_;
+    $c->forward('validateform');
+    $c->forward('get_cert');
+    $c->forward('cert_p12');
+    $c->forward('export_fingerprint');
+}
+
+sub validateform : Private {
+    my ($self,$c,) = @_;
+    $c->stash(
+        service => $c->request->param('service'),
+        certificate_cn => $c->request->param('certificate_cn'),
+        certificate_email => $c->request->param('certificate_email'),
+        certificate_pwd => $c->request->param('certificate_pwd'),
+    );
+
+}
+
+sub export_fingerprint : Local {
+    my ($self, $c) = @_;
+    my $stash = $c->stash;
+    my $cwd = $cert_dir;
+    my $pass = $stash->{'certificate_pwd'};
+    my $certfile = $stash->{'certificate_cn'};
+    my $certp12 = Crypt::OpenSSL::PKCS12->new_from_file("$cwd/$certfile.p12");
+    if ($certp12->mac_ok($pass)){
```

Comment:
What happens if this fails, 

In the code we just continue like nothing happened. You're sure this doesn't break your code down the line ?